Quark-Engine integration

pulorsok opened the pull request:
# Quark-Engine integration

### Requirements:

 - Python 3.7+
 - Quark-Engine 21.01.3+


### Add "Quark Analysis" option

Add "Quark Analysis" as one of the "decode quick pick" items.

Then Quark will be executed after the decompilation of Jadx. And the report will appear as a JSON file in the decoded directory.

![](https://i.imgur.com/p13G4tV.gif)


### A warning message pops up when Quark is not installed
When "Quark Analysis" is selected, it will first check whether Quark-Engine has been installed. If not, it will show an error message and exit the process.

![](https://i.imgur.com/OkehC2T.png)


### Show report as WebView

The summary report of the APK will be shown as a WebView at the bottom panel.

You can also right-click the report file to show the WebView (summary report).


![](https://i.imgur.com/fH5oCvk.png)


Righ-click quark report file to show WebView (summary report).

![](https://i.imgur.com/NQTITYa.gif)

### Source Code Navigation

In WebView (summary report), click on the description of potential malicious activities, the editor panel will open the exact .smali file and highlight the whole block of the function where the activities are detected.

Note: 
Meaning of colors in the highlighted smali codes.
- Green: The whole block of function.
- Yellow: Where the native APIs are called. 


![](https://i.imgur.com/u8CpOx8.gif)

Surendrajat commented on src/extension.ts:
```diff
@@ -70,11 +71,28 @@ export function activate(context: vscode.ExtensionContext): void {
+    const quarkReportCommand = vscode.commands.registerCommand(
+        "apklab.quarkReport",
+        (uri: vscode.Uri) => {
+            updateTools()
```
I think this can be removed as `quark` does not depened on these tools for analysis. Simple call it like `apklab.patchApkForHttps`

Surendrajat commented on src/quark-tools.ts:
```diff
@@ -0,0 +1,467 @@
+     */
+    export function checkQuarkInstalled(): boolean {
+        // TODO: Find a better way to check python and quark install.
+        const cmd = `cd ~/ && quark`;
```
It won't work in windows. Maybe just call `quark` instead?

pulorsok replied:
Quark-Engine will write a log file when executed.
To prevent the log file from being written at the root directory, 
I use the command `cd ~/` before `quark`

Not sure if Windows have the same problem.
But I will add a function to check OS platform,
and use a different command for each platform.

Surendrajat replied:
@pulorsok no need to `cd` at all now. `quark` now writes logs in `~/.quark-engine/`

Surendrajat commented on src/quark-tools.ts:
```diff
@@ -0,0 +1,467 @@
+    }
+
+    /**
+     * Analyzing APK useing Quark-Engine.
```
using

Surendrajat commented on src/quark-tools.ts:
```diff
@@ -0,0 +1,467 @@
+            (progress, token) => {
+                return new Promise<void>((resolve) => {
+                    const quark = child_process.exec(script, function (error) {
+                        if (error) {
```
Showing logs in outputchannel can be useful during analyzing. Not sure if quark spits some useful logs.

Surendrajat commented on src/quark-tools.ts:
```diff
@@ -0,0 +1,467 @@
+                        resolve();
+                    });
+                    token.onCancellationRequested(() => {
+                        if (!quark.killed) {
```
It will only be useful if user can somehow cancel the request. You can show a progress notification like decode/decompile process with a cancel button.

Surendrajat commented on src/quark-tools.ts:
```diff
@@ -0,0 +1,467 @@
+            }
+          }
+        }
+        @import 'https://fonts.googleapis.com/css?family=Montserrat:300,400,700';
```
Can we use some non-goolge fonts maybe? :) Not a issue, just in case.

pulorsok replied:
Sure, I will fix it.

Surendrajat reviewed (changes requested):
Hi @pulorsok, thanks for the PR. Looking great overall. Just adding small comments here and there.

Surendrajat commented on src/extension.ts:
```diff
@@ -70,11 +71,28 @@ export function activate(context: vscode.ExtensionContext): void {
+    const quarkReportCommand = vscode.commands.registerCommand(
+        "apklab.quarkReport",
+        (uri: vscode.Uri) => {
+            updateTools()
```
I think this can be removed as `quark` does not depend on these tools for analysis. Simple call it like `apklab.patchApkForHttps`

pulorsok replied:
You're right, 
I didn't notice that, will fix it too.

Surendrajat commented on src/quark-tools.ts:
```diff
@@ -0,0 +1,467 @@
+ * @return The path of source code file.
+ */
+function functionToPath(func: any): string {
+    let srcPath = "/smali/" + func.class + ".smali";
```
It won't work with muti-dex apk where smali code is spit into `smali`, `smali_classes2`... dirs.

amsharma44 reviewed (changes requested):
Thank for the PR @pulorsok. Awesome work, just a minor concern I have.

pulorsok commented:
@Surendrajat @amsharma44 
Thank you for the comments, I will fix these issues ASAP.

And there are still a few works I will do after.
- Prettier quark report style on WebView :
I will make the report WebView more pretty and readable.

- Quark test unit :
The test unit will be finished next week.

- Same malicious activity in multiple smali files :
Sometime, the same malicious activity will be detected in several smali files.
However, source code navigation will open only a file to show where the activity is.
I will figure out a way to fix this issue (maybe add a hoover to let the user choose which file to open).

Surendrajat commented on src/quark-tools.ts:
```diff
@@ -0,0 +1,467 @@
+        apkDecodedDir: string
+    ): Promise<void> {
+        const reportPath = path.join(apkDecodedDir, `quarkReport.json`);
+        const quarkRulePath = path.join(apkDecodedDir, "quark-rules/");
```
After https://github.com/quark-engine/quark-engine/pull/107, this won't work and can be removed. Failure logs: https://del.dog/gijimobife.txt

Surendrajat commented:
@pulorsok please this PR rebase on master.

pulorsok commented:
Hi @Surendrajat ,
I fixed the problem you mentioned and added a test unit for quark analysis.
I also improved the Webview quark report to get a better UI.
Now it can display multiple functions where the activity is located.

Click on the description of the potentially malicious activity, it will expand and show the detailed of the function (green box), including the class name, method name and smali code that executes the API.
Click on the green box and it will navigate to the source code file.

![](https://i.imgur.com/r8lMKM9.gif)

Surendrajat commented:
@pulorsok Thanks for update. Please check the [recent changes on `quark` side](https://github.com/quark-engine/quark-engine/compare/v21.01.3...master) as well.

Surendrajat commented on src/common.ts:
```diff
@@ -23,3 +23,8 @@ export const extensionConfigName = "apklab";
+/**
+ * Quark-Engine data directory for log file and quark-rules.
+ */
+export const quarkDataDirPath = path.join(apklabDataDir, "quarkDataDir");
```
This won't be needed once new `quark-engine` is released.

Surendrajat commented on src/quark-tools.ts:
```diff
@@ -0,0 +1,386 @@
+ * @return The path of the source code file.
+ */
+function functionToPath(srcDir: string, func: any): string {
+    outputChannel.appendLine(`Search smali file: ${func.class}`);
```
```suggestion
    outputChannel.appendLine(`Searching smali file: ${func.class}`);
```

Surendrajat commented on src/quark-tools.ts:
```diff
@@ -0,0 +1,386 @@
+        }
+        const cmd: string = process.platform.startsWith("win")
+            ? `quark`
+            : `cd ${String(quarkDataDirPath)} && quark`;
```
I meant to remove `cd` entirely :)
You will no longer need it anyway as the logs will go in `~/.quark-engine`.

Surendrajat commented on src/quark-tools.ts:
```diff
@@ -0,0 +1,386 @@
+            child_process.execSync(cmd);
+            return true;
+        } catch (error) {
+            outputChannel.appendLine(`Catched error from Quark install check`);
```
```suggestion
            outputChannel.appendLine(`Caught error from Quark install check`);
```

Surendrajat commented on src/quark-tools.ts:
```diff
@@ -0,0 +1,386 @@
+        apkDecodedDir: string
+    ): Promise<void> {
+        const reportPath = path.join(apkDecodedDir, `quarkReport.json`);
+        const quarkRulePath = path.join(quarkDataDirPath, "quark-rules/");
```
This can be safely removed now :)

Surendrajat commented on src/quark-tools.ts:
```diff
@@ -0,0 +1,386 @@
+        const reportPath = path.join(apkDecodedDir, `quarkReport.json`);
+        const quarkRulePath = path.join(quarkDataDirPath, "quark-rules/");
+
+        const script = `cd ${quarkDataDirPath} && quark -a ${apkFilePath} -r ${quarkRulePath} -o ${reportPath}`;
```
Same here. Since logs and rules will go in `~/.quark-engine` by default we don't need to do `cd` or to pass `-r`.
```suggestion
        const script = `quark -a ${apkFilePath} -o ${reportPath}`;
```

Surendrajat commented on src/quark_html.ts:
```diff
@@ -0,0 +1,304 @@
+        </tr>`;
+
+        for (const pkey in report[key]["api_call"]) {
+            // acitivity = report[pkey.split("-")[0]]["crime"];
```
```suggestion
```

Surendrajat commented on src/test/suite/extension.test.ts:
```diff
@@ -71,11 +71,50 @@ describe("Extension Test Suite", function () {
+        const testApkPath = path.resolve(simpleKeyboardDir, "test.apk");
+        console.log(`Decoding ${testApkPath}...`);
+        await apktool.decodeAPK(testApkPath, [], false, true);
+        const jsonFilePath = path.join(simpleKeyboardDir, "decoded_files.json");
```
No need to check `decoded_files.json` in this test. Please check `qurark_report.json` directly.

Surendrajat reviewed (changes requested):
@pulorsok Leaving some comments here. Mostly related to new `quark-engine` changes.

Surendrajat commented:
@pulorsok could you please rebase it once again? Seems like some conflicts with master again.

Surendrajat commented:
@pulorsok I tested the changes, and it seems we are almost there :)

There are few things that need further changes:
1. The https://github.com/Surendrajat/apklab-testdata submodule needs update to fix the tests.
2. Resulting WebView needs some tweaks. See this:
https://user-images.githubusercontent.com/11215000/105835087-5bfd0e00-5ff1-11eb-846f-565b50490848.mp4

My concerns here are:
- The WebView opens in upper-half and not in lower half which is not intuitive (maybe open it in full).
- The space between `POTENTIAL MALICIOUS ACTIVITIES` and `CONFIDENCE` is a lot.
- The WebView expands out of the TextEditor view (because of above space, I guess).

Surendrajat commented:
@pulorsok I've added the missing step from GitHub workflow to install `quark-egnine` before tests. Please pull the changes and rebase on master once. It should fix the tests.

Surendrajat commented on src/test/suite/extension.test.ts:
```diff
@@ -71,11 +71,42 @@ describe("Extension Test Suite", function () {
     });
 
+    // Quark test unit
+    it("Decode SimpleKeyboard.apk", async function () {
```
```suggestion
    it("Quark-engine Analysis", async function () {
```

Surendrajat commented on src/test/suite/extension.test.ts:
```diff
@@ -71,11 +71,42 @@ describe("Extension Test Suite", function () {
+            assert.fail(`Report data ${reportFile} not found!`);
+        }
+        const reportData = fs.readFileSync(reportFile, "utf-8");
+        console.log("Comparing file list...");
```
```suggestion
        console.log("Comparing analysis report...");
```

Surendrajat reviewed (approved):
@pulorsok Great job!
Looks good overall. I'm planning to move docs to a proper repo, so we can add more .gifs and guide there.
Thanks again for your contribution. Thanks, @krnick, for promptly handling issues with integration and @18z, for reaching out to me :D
